File: Python/board.py
from enum import Enum

class Turn(Enum):
    BLACK       = 1
    WHITE       = 2

class State(Enum):
    BLACK       = 1
    WHITE       = 2
    POSSIBLE    = 3
    NONE        = 0

class Board:
    width       = 8
    height      = 8

    def __init__(self):
        pass

    def init(self):
        self.state = [[State.NONE for _ in range(self.width)] for _ in range(self.height)]
        self.count = [[0 for _ in range(self.width)] for _ in range(self.height)]

        self.state[3][3] = State.BLACK
        self.state[3][4] = State.WHITE
        self.state[4][3] = State.WHITE
        self.state[4][4] = State.BLACK

        self.turn = Turn.WHITE
        self.black = 0
        self.white = 0
        self.end = False

        self.nextTurn()

    def nextTurn(self):
        self.turn = Turn.BLACK if self.turn == Turn.WHITE else Turn.WHITE

        self.black = 0
        self.white = 0
        for y in range(self.height):
            for x in range(self.width):
                if self.state[y][x] == State.BLACK: self.black += 1
                elif self.state[y][x] == State.WHITE: self.white += 1

        possible = False
        for y in range(self.height):
            for x in range(self.width):
                self.count[y][x] = self.reverse(x, y, False)
                if self.count[y][x] > 0:
                    self.state[y][x] = State.POSSIBLE
                    possible = True

        return possible

    def reverse(self, x, y, execute = True):
        if self.state[y][x] == State.BLACK or self.state[y][x] == State.WHITE:
            return 0

        self.state[y][x] = State.NONE if execute != True else (State.BLACK if self.turn == Turn.BLACK else State.WHITE)

        delta = ((0, 1), (1, 1), (1, 0), (1, -1), (0, -1), (-1, -1), (-1, 0), (-1, 1))

        count = 0
        for d in delta:
            count += self.reverseDiagonal(x, y, d[0], d[1], execute)

        return count

    def reverseDiagonal(self, x, y, dx, dy, execute = True):
        if self.turn == Turn.BLACK:
            mine = State.BLACK
            other = State.WHITE
        else:
            mine = State.WHITE
            other = State.BLACK

        count = 0
        while True:
            x += dx
            y += dy

            if x < 0 or x >= self.width or y < 0 or y >= self.height:
                break

            if self.state[y][x] == other:
                count += 1
            elif self.state[y][x] == mine:
                if execute:
                    while True:
                        x -= dx
                        y -= dy
                        if x < 0 or x >= self.width or y < 0 or y >= self.height:
                            break

                        if self.state[y][x] == other:
                            self.state[y][x] = mine
                        else:
                            break
                return count
            else:
                break
        return 0

    def putStone(self, x, y):
        if self.state[y][x] == State.POSSIBLE:
            self.reverse(x, y)

            if self.nextTurn() == False:
                if self.nextTurn() == False:
                    self.end = True

            return True

        return False

File: C++/Board.cpp
#include "Board.h"

Board::Board() : turn(Turn::WHITE), black(0), white(0), end(false)
{
}

void Board::init()
{
	for (int y = 0; y < height; ++y)
		for (int x = 0; x < width; ++x) {
			state[y][x] = State::NONE;
			count[y][x] = 0;
		}

	state[3][3] = State::BLACK;
	state[3][4] = State::WHITE;
	state[4][3] = State::WHITE;
	state[4][4] = State::BLACK;

	turn = Turn::WHITE;
	black = white = 0;
	end = false;

	nextTurn();
}

bool Board::nextTurn()
{
	// toggle turn
	turn = (turn == Turn::BLACK ? Turn::WHITE : Turn::BLACK);

	// update score
	white = 0;
	black = 0;
	for (int y = 0; y < height; ++y)
		for (int x = 0; x < width; ++x)
			if (state[y][x] == State::BLACK) black++;
			else if (state[y][x] == State::WHITE) white++;

	// update possible & count
	bool possible = false;
	for (int y = 0; y < height; ++y)
		for (int x = 0; x < width; ++x) {
			count[y][x] = reverse(x, y, false);
			if (count[y][x] > 0) {
				state[y][x] = State::POSSIBLE;
				possible = true;
			}
		}
	return possible;
}

int Board::reverse(int x, int y, bool execute)
{
	if (state[y][x] == State::BLACK || state[y][x] == State::WHITE) {
		return 0;
	}

	state[y][x] = (execute ? (turn == Turn::BLACK ? State::BLACK : State::WHITE) : State::NONE);

	static const int delta[8][2] = {
		{ 0, 1 },
		{ 1, 1 },
		{ 1, 0 },
		{ 1, -1 },
		{ 0, -1 },
		{ -1, -1 },
		{ -1, 0 },
		{ -1, 1 },
	};

	int count = 0;
	for (auto d : delta) {
		count += reverse(x, y, d[0], d[1], execute);
	}
	return count;
}

int Board::reverse(int x, int y, int dx, int dy, bool execute)
{
	State self = turn == Turn::BLACK ? State::BLACK : State::WHITE;
	State other = turn == Turn::BLACK ? State::WHITE : State::BLACK;

	int count = 0;
	for (x += dx, y += dy; x >= 0 && x < width && y >= 0 && y < height; x += dx, y += dy) {
		if (state[y][x] == other) {
			count++;
		}
		else if (state[y][x] == self) {
			if (execute) {
				for (x -= dx, y -= dy; x >= 0 && x < 8 && y >= 0 && y < 8; x -= dx, y -= dy) {
					if (state[y][x] == other) state[y][x] = self;
					else break;
				}
			}
			return count;
		}
		else {
			return 0;
		}
	}

	return 0;
}

bool Board::putStone(int x, int y)
{
	if (state[y][x] == State::POSSIBLE) {
		reverse(x, y);
		if (nextTurn() == false) {
			if (nextTurn() == false) {
				end = true;
			}
		}

		return true;
	}

	return false;
}

void Board::getScore(int& b, int& w) const
{
	b = black;
	w = white;
}